File: leetcode_python/Array/next-permutation.py
"""

31. Next Permutation
Medium

Implement next permutation, which rearranges numbers into the lexicographically next greater permutation of numbers.

If such an arrangement is not possible, it must rearrange it as the lowest possible order (i.e., sorted in ascending order).

The replacement must be in place and use only constant extra memory.

 
Example 1:

Input: nums = [1,2,3]
Output: [1,3,2]

Example 2:

Input: nums = [3,2,1]
Output: [1,2,3]

Example 3:

Input: nums = [1,1,5]
Output: [1,5,1]

Example 4:

Input: nums = [1]
Output: [1]
 

Constraints:

1 <= nums.length <= 100
0 <= nums[i] <= 100


"""

# V0
class Solution:
    def nextPermutation(self, nums):
        i = j = len(nums)-1
        while i > 0 and nums[i-1] >= nums[i]:
            i -= 1
        if i == 0:   # nums are in descending order
            nums.reverse()
            return 
        k = i - 1    # find the last "ascending" position
        while nums[j] <= nums[k]:
            j -= 1
        nums[k], nums[j] = nums[j], nums[k]  
        l, r = k+1, len(nums)-1  # reverse the second part
        while l < r:
            nums[l], nums[r] = nums[r], nums[l]
            l +=1
            r -= 1

# V0'
class Solution(object):
    def nextPermutation(self, num):
        k, l = -1, 0
        for i in range(len(num) - 1):
            if num[i] < num[i + 1]:
                k = i

        if k == -1:
            num.reverse()
            return

        for i in range(k + 1, len(num)):
            if num[i] > num[k]:
                l = i
        num[k], num[l] = num[l], num[k]
        num[k + 1:] = num[:k:-1] ### dounle check here ###

# V1
# IDEA : 2 POINTERS
# https://leetcode.com/problems/next-permutation/discuss/14054/Python-solution-with-comments.
# https://leetcode.com/problems/next-permutation/discuss/545263/Python-3-Easy-to-understand
class Solution:
    def nextPermutation(self, nums):
        i = j = len(nums)-1
        while i > 0 and nums[i-1] >= nums[i]:
            i -= 1
        if i == 0:   # nums are in descending order
            nums.reverse()
            return 
        k = i - 1    # find the last "ascending" position
        while nums[j] <= nums[k]:
            j -= 1
        nums[k], nums[j] = nums[j], nums[k]  
        l, r = k+1, len(nums)-1  # reverse the second part
        while l < r:
            nums[l], nums[r] = nums[r], nums[l]
            l +=1
            r -= 1

# V1
# https://leetcode.com/problems/next-permutation/discuss/229211/Python-solution
# IDEA :
# First observe that if a list of numbers is in descending order, then there is no lexicographically next greater permutation. Hence for i in range(n-1,0,-1), we search for the first occurrence of i such that nums[i] < nums[i+1]. If no such i exists, the list is in descending order, and we use nums.reverse() to reverse the list in-place. Otherwise, if such i exists, then nums[i-1] will be updated to get the lexicographically next greater permutation. Next, we need to search for the smallest number in nums[i:] that's larger than nums[i-1], and swap it with nums[i-1]. Note that nums[i:] is sorted in descending order. Hence we start with j = i, and while j < n and nums[j] > nums[i-1], we do idx = j, j += 1. When we are out of the while loop, nums[idx] will be the smallest number in nums[i:] that's larger than nums[i]. We then swap nums[idx] and nums[i-1]. After the swap, we just need to sort nums[i:] in ascending order to get the lexicographically next greater permutation. This can be achieved fairly easily in-place, because nums[i:] is already in descending order, and we just need to invert nums[i:] in-place to sort nums[i:] in ascending order.
# To illustrate the algorithm with an example, consider nums = [2,3,1,5,4,2]. It is easy to see that i = 2 is the first i (from the right) such that nums[i] < nums[i+1]. Then we swap nums[2] = 1 with the smallest number in nums[3:] that is larger than 1, which is nums[5] = 2, after which we get nums = [2,3,2,5,4,1]. To get the lexicographically next greater permutation of nums, we just need to sort nums[3:] = [5,4,1] in-place. Finally, we reach nums = [2,3,2,1,4,5].
# Time complexity: O(n), space complexity: O(1).
class Solution(object):
    def nextPermutation(self, nums):
        """
        :type nums: List[int]
        :rtype: void Do not return anything, modify nums in-place instead.
        """
        n = len(nums)
        for i in range(n-1, 0, -1):
            if nums[i] > nums[i-1]:
                j = i
                while j < n and nums[j] > nums[i-1]:
                    idx = j
                    j += 1
                nums[idx], nums[i-1] = nums[i-1], nums[idx]
                for k in range((n-i)//2):
                    nums[i+k], nums[n-1-k] = nums[n-1-k], nums[i+k]
                break
        else:
            nums.reverse()

# V1 
# https://zxi.mytechroad.com/blog/algorithms/array/leetcode-31-next-permutation/
# VIDEO DEMO 
# https://www.youtube.com/watch?v=1ja5s9TmwZM
class Solution:
  def nextPermutation(self, nums):
    n = len(nums)
    i = n - 2
    while i >= 0 and nums[i] >= nums[i + 1]: 
        i -= 1
    if i >= 0: # if can find such "i" fit what problem need 
      j = n - 1
      while j >= 0 and nums[j] <= nums[i]: 
        j -= 1
      nums[i], nums[j] = nums[j], nums[i]
    # reverse
    nums[i+1:] = nums[i+1:][::-1]

# V1' 
# http://bookshadow.com/weblog/2016/09/09/leetcode-next-permutation/
class Solution(object):
    def nextPermutation(self, nums):
        """
        :type nums: List[int]
        :rtype: void Do not return anything, modify nums in-place instead.
        """
        size = len(nums)
        for x in range(size - 1, -1, -1):
            if nums[x - 1] < nums[x]:
                break
        if x > 0:
            for y in range(size - 1, -1, -1):
                if nums[y] > nums[x - 1]:
                    nums[x - 1], nums[y] = nums[y], nums[x - 1]
                    break
        for z in range((size - x) / 2):
            nums[x + z], nums[size - z - 1] = nums[size - z - 1], nums[x + z]

# V1
# https://leetcode.com/problems/next-permutation/discuss/162049/Explanations-(Java-Python)
class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """      
        first_inc = len(nums) - 2
        
        while first_inc >= 0 and nums[first_inc] >= nums[first_inc + 1]:
            first_inc -= 1
    
        if first_inc != -1:  # nums is not non-increasing as a whole
            for i in range(len(nums) - 1, first_inc, -1):
                if nums[i] > nums[first_inc]:
                    nums[first_inc], nums[i] = nums[i], nums[first_inc]
                    break
    
        nums[first_inc + 1: len(nums)] = nums[first_inc + 1: len(nums)][::-1]

# V1'' 
# https://blog.csdn.net/fuxuemingzhu/article/details/82113409
# DMEO
# STEP 1) given 12431
# STEP 2) 12 431 (431 is the decreasing substring) (start from the right hand side of orignal list)
# STEP 3) 12 134 (reverse 431 -> 134)
# STEP 4) 13 124 (find the 1st elment in 134 that bigger than 2)
# STEP 5) answer = 13124
class Solution(object):
    def nextPermutation(self, nums):
        """
        :type nums: List[int]
        :rtype: void Do not return anything, modify nums in-place instead.
        """
        n = len(nums)
        i = n - 1
        while i > 0 and nums[i] <= nums[i - 1]:
            i -= 1
        self.reverse(nums, i, n - 1)
        if i > 0:
            for j in range(i, n):
                if nums[j] > nums[i-1]:
                    self.swap(nums, i-1, j)
                    break
        
    def reverse(self, nums, i, j):
        """
        contains i and j.
        demo :
            nums = [1,2,3,4,5,6]
            i, j = 1,3 
            r = reverse(nums, i, j)
            -> r = [1, 4, 3, 2, 5, 6]
        """
        for k in range(i, (i + j) / 2 + 1):
            self.swap(nums, k, i + j - k)

        
    def swap(self, nums, i, j):
        """
        contains i and j.
        """
        nums[i], nums[j] = nums[j], nums[i]

# V1'''
# https://www.jiuzhang.com/solution/next-permutation/#tag-highlight-lang-python
class Solution:
    # @param num :  a list of integer
    # @return : a list of integer
    def nextPermutation(self, num):
        # write your code here
        for i in range(len(num)-2, -1, -1):
            if num[i] < num[i+1]:
                break
        else:
            num.reverse()
            return num
        for j in range(len(num)-1, i, -1):
            if num[j] > num[i]:
                num[i], num[j] = num[j], num[i]
                break
        for j in range(0, (len(num) - i)//2):
            num[i+j+1], num[len(num)-j-1] = num[len(num)-j-1], num[i+j+1]
        return num

# V1
# IDEA : Single Pass Approach
# https://leetcode.com/problems/next-permutation/solution/
# JAVA
# public class Solution {
#     public void nextPermutation(int[] nums) {
#         int i = nums.length - 2;
#         while (i >= 0 && nums[i + 1] <= nums[i]) {
#             i--;
#         }
#         if (i >= 0) {
#             int j = nums.length - 1;
#             while (nums[j] <= nums[i]) {
#                 j--;
#             }
#             swap(nums, i, j);
#         }
#         reverse(nums, i + 1);
#     }
#
#     private void reverse(int[] nums, int start) {
#         int i = start, j = nums.length - 1;
#         while (i < j) {
#             swap(nums, i, j);
#             i++;
#             j--;
#         }
#     }
#
#     private void swap(int[] nums, int i, int j) {
#         int temp = nums[i];
#         nums[i] = nums[j];
#         nums[j] = temp;
#     }
# }

# V2 
# Time:  O(n)
# Space: O(1)
class Solution(object):
    # @param {integer[]} nums
    # @return {void} Do not return anything, modify nums in-place instead.
    def nextPermutation(self, num):
        k, l = -1, 0
        for i in range(len(num) - 1):
            if num[i] < num[i + 1]:
                k = i

        if k == -1:
            num.reverse()
            return

        for i in range(k + 1, len(num)):
            if num[i] > num[k]:
                l = i
        num[k], num[l] = num[l], num[k]
        num[k + 1:] = num[:k:-1]

File: leetcode_java/src/main/java/LeetCodeJava/Array/NextPermutation.java
package LeetCodeJava.Array;

// https://leetcode.com/problems/next-permutation/description/
/**
 * 31. Next Permutation
 * Solved
 * Medium
 * Topics
 * Companies
 * A permutation of an array of integers is an arrangement of its members into a sequence or linear order.
 *
 * For example, for arr = [1,2,3], the following are all the permutations of arr: [1,2,3], [1,3,2], [2, 1, 3], [2, 3, 1], [3,1,2], [3,2,1].
 * The next permutation of an array of integers is the next lexicographically greater permutation of its integer. More formally, if all the permutations of the array are sorted in one container according to their lexicographical order, then the next permutation of that array is the permutation that follows it in the sorted container. If such arrangement is not possible, the array must be rearranged as the lowest possible order (i.e., sorted in ascending order).
 *
 * For example, the next permutation of arr = [1,2,3] is [1,3,2].
 * Similarly, the next permutation of arr = [2,3,1] is [3,1,2].
 * While the next permutation of arr = [3,2,1] is [1,2,3] because [3,2,1] does not have a lexicographical larger rearrangement.
 * Given an array of integers nums, find the next permutation of nums.
 *
 * The replacement must be in place and use only constant extra memory.
 *
 *
 *
 * Example 1:
 *
 * Input: nums = [1,2,3]
 * Output: [1,3,2]
 * Example 2:
 *
 * Input: nums = [3,2,1]
 * Output: [1,2,3]
 * Example 3:
 *
 * Input: nums = [1,1,5]
 * Output: [1,5,1]
 *
 *
 * Constraints:
 *
 * 1 <= nums.length <= 100
 * 0 <= nums[i] <= 100
 *
 */
public class NextPermutation {

    // V0
    // IDEA : 2 POINTERS (gpt)
    // https://github.com/yennanliu/CS_basics/blob/master/leetcode_python/Array/next-permutation.py#L43
    public void nextPermutation(int[] nums) {

        int i = nums.length - 1;
        int j = nums.length - 1;

        // Find the first decreasing element from the end
        while (i > 0 && nums[i - 1] >= nums[i]) {
            i--;
        }

        // If the entire array is in descending order, reverse it
        if (i == 0) {
            reverse_(nums, 0, nums.length - 1);
            return;
        }

        int k = i - 1; // Find the last "ascending" position

        // Find the element just larger than nums[k]
        while (nums[j] <= nums[k]) {
            j--;
        }

        // Swap the elements at positions k and j
        swap_(nums, k, j);

        // Reverse the sequence from k+1 to the end
        reverse_(nums, k + 1, nums.length - 1);
    }

    // Helper method to swap elements at positions i and j
    private void swap_(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }

    // Helper method to reverse elements from start to end
    private void reverse_(int[] nums, int start, int end) {
        while (start < end) {
            swap_(nums, start, end);
            start++;
            end--;
        }
    }

    // V1
    // IDEA : 2 POINTERS
    // https://leetcode.com/problems/next-permutation/solutions/3473399/beats-100-full-explanation-in-steps/
    public void nextPermutation_1(int[] nums) {
        int ind1=-1;
        int ind2=-1;
        // step 1 find breaking point
        for(int i=nums.length-2;i>=0;i--){
            if(nums[i]<nums[i+1]){
                ind1=i;
                break;
            }
        }
        // if there is no breaking  point
        if(ind1==-1){
            reverse(nums,0);
        }

        else{
            // step 2 find next greater element and swap with ind2
            for(int i=nums.length-1;i>=0;i--){
                if(nums[i]>nums[ind1]){
                    ind2=i;
                    break;
                }
            }

            swap(nums,ind1,ind2);
            // step 3 reverse the rest right half
            reverse(nums,ind1+1);
        }
    }
    void swap(int[] nums,int i,int j){
        int temp=nums[i];
        nums[i]=nums[j];
        nums[j]=temp;
    }
    void reverse(int[] nums,int start){
        int i=start;
        int j=nums.length-1;
        while(i<j){
            swap(nums,i,j);
            i++;
            j--;
        }
    }

    // V2
    // IDEA : 2 POINTERS
    // https://leetcode.com/problems/next-permutation/solutions/5205798/java-solution-0ms-solution-100-faster-solution-full-explanation-code/
    public void nextPermutation_2(int[] nums) {
        // Traverse form the end and find the element which is smaller
        // than its next element. Keep that index as ind

        // Traverse the array from the back and find the element which is
        // nearest element grater that the element present in index ind.
        // Swap them.

        // Reverse the array after the index the index ind

        int ind = -1;

        int n = nums.length;

        for(int i = n - 1; i >= 1; i--) {
            if(nums[i] > nums[i - 1]) {
                ind = i - 1;
                break;
            }
        }

        if(ind == -1) {
            int i = 0, j = n - 1;

            while(i < j) {
                int temp = nums[i];
                nums[i] = nums[j];
                nums[j] = temp;

                i++;
                j--;
            }

            return;
        }
        else {
            for(int i = n - 1; i > ind; i--) {
                if(nums[i] > nums[ind]) {
                    int temp = nums[ind];
                    nums[ind] = nums[i];
                    nums[i] = temp;
                    break;
                }
            }
        }

        int i = ind + 1, j = n - 1;

        while(i < j) {
            int temp = nums[i];
            nums[i] = nums[j];
            nums[j] = temp;

            i++;
            j--;
        }

        return;
    }

    // V3
    // IDEA : 2 POINTERS (gpt)
    /**
     *  IDEA :
     *
     *  Explanation of the Algorithm:
     *
     * 	1.	Identify the first decreasing element from the end:
     * 	    •	Traverse the array from the right and find the first index i such that nums[i] < nums[i + 1].
     * 	    •	This is the pivot point where the next permutation is possible.
     *
     * 	2.	Find the smallest element larger than nums[i] to the right of i:
     * 	    •	From the end of the array, find the first index j such that nums[j] > nums[i].
     *
     * 	3.	Swap nums[i] and nums[j]:
     * 	    •	Swap these elements to ensure the next permutation is lexicographically greater.
     *
     * 	4.	Reverse the suffix starting from i + 1:
     * 	    •	Reverse the elements to the right of i to ensure they are in ascending order, giving the smallest lexicographical order for the suffix.
     *
     * 	5.	If no i is found (i.e., the array is in descending order):
     * 	    •	Reverse the entire array to reset it to the smallest permutation.
     *
     */
    public void nextPermutation_3(int[] nums) {
        int n = nums.length;
        /**
         *  why set i = n - 2 ?
         *
         *  The initialization of int i = n - 2; in the next permutation algorithm is essential for the logic of finding the first decreasing element when traversing the array from right to left. Here’s why it starts at n - 2:
         *
         * Understanding the Reason:
         *
         * 	1.	i Represents the Pivot Point:
         * 	    •	The algorithm tries to find the largest index i where nums[i] < nums[i + 1].
         * 	    •	This index i is called the pivot because it is where the current permutation can be adjusted to create the next lexicographically greater permutation.
         * 	2.	Start from the Second-to-Last Element (n - 2):
         * 	    •	The last element (nums[n - 1]) has no element to its right, so it cannot be compared with anything.
         * 	    •	To compare nums[i] with nums[i + 1], i must start from n - 2, the second-to-last index.
         * 	3.	Traverse from Right to Left:
         * 	    •	By moving from right to left, the algorithm ensures that we find the last occurrence of nums[i] < nums[i + 1], which is the correct pivot point for generating the next permutation.
         * 	    •	If no such i is found (i.e., the array is in descending order), the array is reversed to reset it to the smallest permutation.
         *
         */
        int i = n - 2;

        // Step 1: Find the first decreasing element from the end
        while (i >= 0 && nums[i] >= nums[i + 1]) {
            i--;
        }

        if (i >= 0) { // If such an element is found
            // Step 2: Find the smallest element larger than nums[i]
            int j = n - 1;
            while (nums[j] <= nums[i]) {
                j--;
            }

            // Step 3: Swap nums[i] and nums[j]
            swap_3(nums, i, j);
        }

        // Step 4: Reverse the suffix starting at i + 1
        reverse_3(nums, i + 1, n - 1);
    }

    // Helper method to swap two elements
    private void swap_3(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }

    // Helper method to reverse part of the array
    private void reverse_3(int[] nums, int start, int end) {
        while (start < end) {
            swap_3(nums, start, end);
            start++;
            end--;
        }
    }

}
